File: src/engines/net_info/net_info.rs
use crate::iface::{Iface, SysInfo};
use crate::engines::NetInfoArgs;



#[derive(Default)]
pub struct NetworkInfo {
    iface       : Iface,
    state       : String, 
    if_type     : String, 
    mac         : String, 
    ip          : String, 
    cidr        : String, 
    host_len    : String, 
    mtu         : String, 
    gateway_mac : String, 
    gateway_ip  : String,
    broadcast   : String,
}



impl NetworkInfo {

    pub fn new(_args: NetInfoArgs) -> Self {
        Self { ..Default::default() }
    }



    pub fn execute(&mut self) {
        SysInfo::ifaces()
            .into_iter()
            .enumerate()
            .for_each(|(i, iface)|{
                self.set_iface(iface)
                    .set_state()
                    .set_type()
                    .set_mac()
                    .set_ip()
                    .set_cidr()
                    .set_len_host()
                    .set_mtu()
                    .set_gateway_mac()
                    .set_gateway_ip()
                    .set_broadcast()
                    .display_info(i);
            }
        )
    }



    fn set_iface(&mut self, iface: String) -> &mut Self {
        self.iface = Iface::new(&iface);
        self
    }



    fn set_state(&mut self) -> &mut Self {
        self.state = self.iface.state().unwrap_or_else(|_| "Unknown".to_string());
        self
    }

    
    
    fn set_type(&mut self) -> &mut Self {        
        self.if_type = self.iface.if_type();
        self
    }
    


    fn set_mac(&mut self) -> &mut Self {
        self.mac = match self.iface.mac() {
            Ok(mac) => mac.to_string(),
            Err(_)  => "Unknown".to_string(),
        };

        self
    }

    
    
    fn set_ip(&mut self) -> &mut Self {
        self.ip = match self.iface.ip() {
            Ok(ip) => ip.to_string(),
            Err(_) => "None".to_string(),
        };

        self
    }

    
    
    fn set_cidr(&mut self) -> &mut Self {
        self.cidr = match self.iface.cidr() {
            Ok(cidr) => cidr,
            Err(_)   => "Unknown".to_string(),
        };

        self
    }

    
    
    fn set_len_host(&mut self) -> &mut Self {
        let parts: Vec<&str> = self.cidr.split('/').collect();
        if parts.len() != 2 {
            self.host_len = "None".to_string();
            return self;
        }

        let cidr_value: u8 = match parts[1].parse() {
            Ok(value) => value,
            Err(_)    => {
                self.host_len = "Unknown".to_string();
                return self;
            }
        };

        if cidr_value > 32 {
            self.host_len = "None".to_string();
            return self;
        }

        let host_bits   = 32 - cidr_value;
        let total_hosts = 2u32.pow(host_bits as u32);

        if cidr_value >= 31 {
            self.host_len = total_hosts.to_string();
            return self;
        }
        
        self.host_len = (total_hosts - 2).to_string();

        self
    }



    fn set_mtu(&mut self) -> &mut Self {
        self.mtu = self.iface.mtu().unwrap_or_else(|_| "None".to_string());
        self
    }



    fn set_gateway_mac(&mut self) -> &mut Self {
        self.gateway_mac = match self.iface.gateway_mac() {
            Ok(mac) => mac.to_string(),
            Err(_)  => "Unknown".to_string()
        };
        
        self
    }



    fn set_gateway_ip(&mut self) -> &mut Self {
        self.gateway_ip = match self.iface.gateway_ip() {
            Ok(ip) => ip.to_string(),
            Err(_) => "Unknown".to_string()
        };
        
        self
    }



    fn set_broadcast(&mut self) -> &mut Self {
        if &self.if_type == "Loopback" {
            self.broadcast = "None".to_string();
            return self
        }
        
        self.broadcast = match self.iface.broadcast_ip() {
            Ok(ip) => ip.to_string(),
            Err(_) => "Unknown".to_string(),
        };

        self
    }

    
    
    fn display_info(&self, index: usize) {
        println!("#{} Interface: {} - State: {}", index, self.iface.name(), self.state);
        println!("  - Type.......: {}", self.if_type);
        println!("  - MAC........: {}", self.mac);
        println!("  - IP.........: {}", self.ip);
        println!("  - Net Addr...: {}", self.cidr);
        println!("  - Len hosts..: {}", self.host_len);
        println!("  - MTU........: {}", self.mtu);
        println!("  - Gateway MAC: {}", self.gateway_mac);
        println!("  - Gateway IP.: {}", self.gateway_ip);
        println!("  - Broadcast..: {}", self.broadcast);
        println!("")
    }

}



impl crate::EngineTrait for NetworkInfo {
    type Args = NetInfoArgs;
    
    fn new(args: Self::Args) -> Self {
        NetworkInfo::new(args)
    }
    
    fn execute(&mut self) {
        self.execute();
    }
}

File: engines/netinfo/net_info.go
/*
 * Copyright (C) 2025 Oliver R. Calazans Jeronimo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org>.
 */

package netinfo

import (
	"fmt"
	"net"
	"strconv"
	"strings"

	"offscan/internal/conv"
	"offscan/internal/ifaceinfo"
	"offscan/internal/sysinfo"
)



func Run(args []string) {
    newNetInfo(args).execute()
}



type networkInfo struct {
    ifaceList   []net.Interface
    current     *net.Interface
    state       string
    ifType      string
    mac         string
    ip          string
    cidr        string
    hostLen     string
    mtu         string
    gatewayMac  string
    gatewayIP   string
    broadcast   string
}



func newNetInfo(argList []string) *networkInfo {
	args := ParseNetInfoArgs(argList)
    var ifaceList []net.Interface

    if  args.Iface == "" {
        ifaceList = sysinfo.MustAllIfaces()
    } else {
        ifaceList = append(ifaceList, *conv.MustGetIface(args.Iface))
    }

    return &networkInfo{
        ifaceList: ifaceList,
    }
}



func (ni *networkInfo) execute() {
    for idx, iface := range ni.ifaceList {
		ni.current = &iface
        
		ni.setState()
        ni.setType()
        ni.setMAC()
        ni.setIP()
        ni.setCIDR()
        ni.setHostLen()
        ni.setMTU()
        ni.setGatewayMAC()
        ni.setGatewayIP()
        ni.setBroadcast()
        ni.displayInfo(idx)
    }
}



func (ni *networkInfo) setState() {
    state, err := ifaceinfo.State(ni.current)
    
	if err != nil {
        ni.state = "Unknown"
    } else {
        ni.state = state
    }
}



func (ni *networkInfo) setType() {
    ni.ifType = ifaceinfo.Type(ni.current)
}



func (ni *networkInfo) setMAC() {
    ni.mac = ni.current.HardwareAddr.String()    
}



func (ni *networkInfo) setIP() {
    ip, err := ifaceinfo.IPv4(ni.current)
    
	if err != nil {
        ni.ip = "None"
    } else {
        ni.ip = ip.String()
    }
}



func (ni *networkInfo) setCIDR() {
    cidr, err := ifaceinfo.CIDR(ni.current)
    
	if err != nil {
        ni.cidr = "Unknown"
    } else {
        ni.cidr = cidr
    }
}



func (ni *networkInfo) setHostLen() {
    parts := strings.Split(ni.cidr, "/")
    
	if len(parts) < 2 {
        ni.hostLen = "None"
        return
    }

    cidrVal, err := strconv.Atoi(parts[1])
    if err != nil || cidrVal > 32 {
        ni.hostLen = "Unknown"
        return
    }

    hostBits   := 32 - cidrVal
    totalHosts := 1 << uint(hostBits)

    if cidrVal >= 31 {
        ni.hostLen = strconv.Itoa(totalHosts)
    } else {
        ni.hostLen = strconv.Itoa(totalHosts - 2)
    }
}



func (ni *networkInfo) setMTU() {
    ni.mtu = fmt.Sprintf("%d", ni.current.MTU)
}



func (ni *networkInfo) setGatewayMAC() {
    mac, err := ifaceinfo.GatewayMAC(ni.current)

    if err != nil {
        ni.gatewayMac = "Unknown"
    } else {
        ni.gatewayMac = mac.String()
    }
}



func (ni *networkInfo) setGatewayIP() {
    ip, err := ifaceinfo.GatewayIP(ni.current)
    
	if err != nil {
        ni.gatewayIP = "Unknown"
    } else {
        ni.gatewayIP = ip.String()
    }
}



func (ni *networkInfo) setBroadcast() {
    if ni.ifType == "Loopback" {
        ni.broadcast = "None"
        return
    }

	ip, err := ifaceinfo.BroadcastFromCIDR(ni.cidr)
    
	if err != nil {
        ni.broadcast = "Unknown"
    } else {
        ni.broadcast = ip.String()
    }
}



func (ni *networkInfo) displayInfo(index int) {
    fmt.Printf("#%d Interface: %s - State: %s\n", index, ni.current.Name, ni.state)
    fmt.Println("  - Type.......:", ni.ifType)
    fmt.Println("  - MAC........:", ni.mac)
    fmt.Println("  - IP.........:", ni.ip)
    fmt.Println("  - CIDR.......:", ni.cidr)
    fmt.Println("  - Len hosts..:", ni.hostLen)
    fmt.Println("  - MTU........:", ni.mtu)
    fmt.Println("  - Gateway MAC:", ni.gatewayMac)
    fmt.Println("  - Gateway IP.:", ni.gatewayIP)
    fmt.Println("  - Broadcast..:", ni.broadcast)
    fmt.Println()
}